### PR TITLE
Add Desired State for All SubResources for Rabbitmq Controller

### PR DESCRIPTION
**Pull request by euclidgame**

### CODE REVIEW

File: src/controller_examples/rabbitmq_controller/proof/liveness/property.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+{
+    always(lift_state(RMQCluster::desired_state_is(rabbitmq))).leads_to(always(lift_state(current_state_matches(rabbitmq))))
+}
+
+pub open spec fn current_state_matches(rabbitmq: RabbitmqClusterView) -> StatePred<RMQCluster> {
+    |s: RMQCluster| {
+        forall |kind: ResourceKind| {
+            let key = (#[trigger] get_request_view(kind, rabbitmq)).key;
+            &&& s.resources().contains_key(key)
+            &&& resource_state_matches(kind, rabbitmq, s.resources()[key])
```
Comment (euclidgame) [inline src/controller_examples/rabbitmq_controller/proof/liveness/property.rs:41]: @marshtompsxd , I specify the property here. And you can go to `rabbitmq_controller::resource` module for the separate spec for different resources. I guess the one for statefulset may be incorrect.

File: src/controller_examples/rabbitmq_controller/resource/default_user_secret.rs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            })
+        } else {
+            Err(RabbitmqError::Error)
+        }
+    }
+
+    open spec fn resource_state_matches(rabbitmq: RabbitmqClusterView, obj: DynamicObjectView) -> bool {
+        &&& SecretView::unmarshal(obj).is_Ok()
+        &&& SecretView::unmarshal(obj).get_Ok_0().data == resource::make_default_user_secret(rabbitmq).data
+        &&& SecretView::unmarshal(obj).get_Ok_0().type_ == resource::make_default_user_secret(rabbitmq).type_
```
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/resource/default_user_secret.rs:93]: I removed `type_` from Secret in the main branch. You can rebase main and remove this type_ here.

File: src/controller_examples/rabbitmq_controller/resource/headless_service.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                ..state
+            })
+        } else {
+            Err(RabbitmqError::Error)
+        }
+    }
+
+    open spec fn resource_state_matches(rabbitmq: RabbitmqClusterView, obj: DynamicObjectView) -> bool {
+        &&& ServiceView::unmarshal(obj).is_Ok()
+        &&& ServiceView::unmarshal(obj).get_Ok_0().spec == resource::make_headless_service(rabbitmq).spec
```
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/resource/headless_service.rs:92]: We probably don't want to talk about the entire spec of a service, since the clusterIP(s) is not decided by the controller (tho we haven't specified the built-in controller for service yet).

File: src/controller_examples/rabbitmq_controller/resource/service.rs [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                ..state
+            })
+        } else {
+            Err(RabbitmqError::Error)
+        }
+    }
+
+    open spec fn resource_state_matches(rabbitmq: RabbitmqClusterView, obj: DynamicObjectView) -> bool {
+        &&& ServiceView::unmarshal(obj).is_Ok()
+        &&& ServiceView::unmarshal(obj).get_Ok_0().spec == resource::make_main_service(rabbitmq).spec
```
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/resource/service.rs:92]: Similar as headless service

File: src/controller_examples/rabbitmq_controller/common/mod.rs
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/common/mod.rs:1]: Minor: can you revert this change?

File: src/controller_examples/rabbitmq_controller/proof/liveness/property.rs (hunk 1 above)
Comment (marshtompsxd) [reply src/controller_examples/rabbitmq_controller/proof/liveness/property.rs:41]: > I guess the one for statefulset may be incorrect.

Why so?

File: src/controller_examples/rabbitmq_controller/proof/resource.rs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        ResourceKind::PluginsConfigMap => PluginsConfigMapBuilder::state_after_create_or_update_view(obj, state),
+        ResourceKind::ServerConfigMap => ServerConfigMapBuilder::state_after_create_or_update_view(obj, state),
+        ResourceKind::ServiceAccount => ServiceAccountBuilder::state_after_create_or_update_view(obj, state),
+        ResourceKind::Role => RoleBuilder::state_after_create_or_update_view(obj, state),
+        ResourceKind::RoleBinding => RoleBindingBuilder::state_after_create_or_update_view(obj, state),
+        ResourceKind::StatefulSet => StatefulSetBuilder::state_after_create_or_update_view(obj, state),
+    }
+}
+
+pub open spec fn resource_state_matches(kind: ResourceKind, rabbitmq: RabbitmqClusterView, obj: DynamicObjectView) -> bool {
```
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/proof/resource.rs:85]: The term `ResourceKind` is a bit confusing since we have a `Kind`. Can you rename it to `SubResource`? And also rename the variable `kind` to `sub_resource?

**Comment by marshtompsxd** [comment]
@euclidgame Just trying to understand it better: I see that you deduplicate a lot of code by reusing `reconcile_helper`, which implements the super common get-then-update-or-create pattern. How do you manage special cases such as saving the configmap's rv into the reconcile state in this way? Can you point me to the code?

File: src/controller_examples/rabbitmq_controller/resource/config_map.rs
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/resource/config_map.rs:1]: What's the rationale behind moving stuff from spec/ and exec/ to this resource/? Note that we later need to count how many exec/spec/proof code for each controller.

Comment (marshtompsxd) [reply src/controller_examples/rabbitmq_controller/resource/config_map.rs:1]: I just noticed that you only moved stuff from spec/ to resource/, but still the same problem

**Comment by marshtompsxd** [comment]
> @euclidgame Just trying to understand it better: I see that you deduplicate a lot of code by reusing `reconcile_helper`, which implements the super common get-then-update-or-create pattern. How do you manage special cases such as saving the configmap's rv into the reconcile state in this way? Can you point me to the code?

nvm. I figured it out

File: src/controller_examples/rabbitmq_controller/resource/config_map.rs
Comment (marshtompsxd) [reply src/controller_examples/rabbitmq_controller/resource/config_map.rs:1]: Anyway let's don't revert it back too early and I will ask Andrea tmr how to count LoC in Verus.

File: src/controller_examples/rabbitmq_controller/spec/resource/stateful_set.rs [hunk 6]
```diff
@@ -56,6 +56,12 @@ impl ResourceBuilder<StatefulSetView> for StatefulSetBuilder {
             Err(RabbitmqError::Error)
         }
     }
+
+    open spec fn resource_state_matches(rabbitmq: RabbitmqClusterView, obj: DynamicObjectView) -> bool {
+        &&& StatefulSetView::unmarshal(obj).is_Ok()
+        &&& exists |rv: StringView|
```
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/spec/resource/stateful_set.rs:62]: We need to make this stronger by saying the rv is the same as the configmap's

File: src/controller_examples/rabbitmq_controller/proof/liveness/property.rs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+{
+    always(lift_state(RMQCluster::desired_state_is(rabbitmq))).leads_to(always(lift_state(current_state_matches(rabbitmq))))
+}
+
+pub open spec fn current_state_matches(rabbitmq: RabbitmqClusterView) -> StatePred<RMQCluster> {
+    |s: RMQCluster| {
+        forall |sub_resource: SubResource| {
+            let key = (#[trigger] get_request(sub_resource, rabbitmq)).key;
+            &&& s.resources().contains_key(key)
+            &&& resource_state_matches(sub_resource, rabbitmq, s.resources()[key])
```
Comment (marshtompsxd) [inline src/controller_examples/rabbitmq_controller/proof/liveness/property.rs:40]: Maybe you can make the resource_state_matches takes the entire s.resoruces() and the key, so that you can talk about dependencies between objects in the property, like the sts has the rv of the configmap as its pod's annotation

File: src/controller_examples/rabbitmq_controller/proof/liveness/property.rs (hunk 1 above)
Comment (marshtompsxd) [reply src/controller_examples/rabbitmq_controller/proof/liveness/property.rs]: now i see

**Comment by marshtompsxd** [comment]
LGTM. Feel free to merge
